set cluster/route/listener/vhost name len based on CLI opt

rshriram opened the pull request:
Automatically set the max allowed length for cluster, route_config_name and listener name based on the max_stat_name_len specified. The name len is max_stat_name_len - 67 (which is the suffix reserved for internally generated stats).

Signed-off-by: Shriram Rajagopalan <shriram@us.ibm.com>

cc @hennna / @ggreenway FYI.

@mattklein123 this is per discussion on Slack.

mattklein123 commented:
Quick pre-review comments: Will need doc changes. Also, we should do this for RDS and LDS also.

ggreenway commented on include/envoy/server/options.h:
```diff
@@ -130,6 +130,11 @@ class Options {
+  /**
+   * @return uint64_t the maximum name length of a cluster/route/listener.
+   */
+  virtual uint64_t maxUserSuppliedNameLength() PURE;
```
I think that it would be better to have this value be derived from the maxStatNameLen, or possibly have that one be derived from this one.  Being able to specify both seems confusing.

mattklein123 replied:
Yup exactly. No more options are needed. Please rework as @ggreenway suggests.

rshriram replied:
Yeah good point. Does it suffice to just derive max cluster name length as max_stat_name_len - reserved_stat_suffix_len (which happens to be 67 at the moment) ? That would make things considerably simple.

mattklein123 replied:
Yup that's exactly what I would do.

htuch commented:
@hennna please review this from the Google side (I will do the final pass from us).

ggreenway commented on source/common/stats/stats_impl.h:
```diff
@@ -103,6 +111,8 @@ struct RawStatData {
 private:
   static const size_t DEFAULT_MAX_NAME_SIZE = 127;
+  // total length must be equal to DEFAULT_MAX_NAME_SIZE
+  static const size_t DEFAULT_MAX_STAT_SUFFIX_LENGTH = 67;
```
rename to MAX_STAT_SUFFIX_LENGTH?  Since it isn't configurable, I don't think of it as a default; it's the only value.

rshriram replied:
done..

rshriram commented:
@ggreenway I too feel that its probably better to have the identity name be the configurable option and derive the other one. @htuch / @mattklein123 LMK if you want to replace the max-stat-name-len with max-user-field-name-len or something.. 

Btw, its not  just cluster. Its cluster name, route_config_name, listener name.

mattklein123 commented:
> Given that we just added the command-line option a couple weeks ago, @mattklein123 what do you think about replacing it with one for max user supplied name len, and computing the max name len from that?

That's fine. Since we haven't release 1.5.0 yet we can change it. Please just make the defaults such that we don't need to do a full restart (the shared memory version should be the same).

rshriram commented:
@mattklein123 / @ggreenway PTAL.. changed the CLI option and compile time option as well.

ggreenway commented on source/common/config/utility.cc:
```diff
@@ -119,5 +123,13 @@ std::string Utility::resourceName(const ProtobufWkt::Any& resource) {
+                                          const std::string& name) {
+  if (name.length() > Stats::RawStatData::maxUserSuppliedNameLength()) {
+    throw EnvoyException(fmt::format("{}: name exceeds allowed maximum {}", error_prefix,
+                                     Stats::RawStatData::maxUserSuppliedNameLength()));
```
Please also print the passed in name, and possibly the length, so it's easier to figure out what was too big and by how much it was too big.

rshriram replied:
I can, but be warned that its of little use. The fields are named "name" which is not of much use. Thats the reason I opted to use the error_prefix and have some meaningful error message as part of the error.

ggreenway commented on source/common/stats/stats_impl.h:
```diff
@@ -103,6 +109,8 @@ struct RawStatData {
 private:
   static const size_t DEFAULT_MAX_NAME_SIZE = 127;
+  // total length must be equal to DEFAULT_MAX_NAME_SIZE
+  static const size_t MAX_STAT_SUFFIX_LENGTH = 67;
```
Please only define this value in one place.  It is duplicated in options_impl.cc right now.

ggreenway commented on source/server/options_impl.cc:
```diff
@@ -144,6 +150,6 @@ OptionsImpl::OptionsImpl(int argc, char** argv, const HotRestartVersionCb& hot_r
   parent_shutdown_time_ = std::chrono::seconds(parent_shutdown_time_s.getValue());
   max_stats_ = max_stats.getValue();
-  max_stat_name_length_ = max_stat_name_len.getValue();
+  max_stat_name_length_ = max_obj_name_len.getValue() + ENVOY_MAX_STAT_SUFFIX_LENGTH;
```
The user-specified option is now max_stat_name_len.  For consistency, please replace the method  Options::maxStatNameLength() with Options::maxObjNameLen()

rshriram replied:
I intentionally left it as it is. Replacing it with maxObjectNameLen will require each call site to use maxObjectNameLen + maxStatSuffixLen. That seems unnecessary detail across the code base where maxStatNameLen is used.

ggreenway replied:
There should only end up being one call-site: RawStatData::configure().  And then there will be 2 static functions on RawStatData; maxNameLength() and maxObjNameLength(), one of which will be implemented in terms of the other (adding or subtracting 67).

ggreenway commented on test/server/hot_restart_impl_test.cc:
```diff
@@ -132,15 +136,15 @@ TEST_P(HotRestartImplAlignmentTest, objectOverlap) {
   for (uint64_t i = 0; i < num_stats_; i++) {
     std::string name = fmt::format("{}zzzzzzzzzzzzzzzzzzzzzzzzzzzz", i)
-                           .substr(0, Stats::RawStatData::maxNameLength());
+                           .substr(0, Stats::RawStatData::maxObjNameLength());
```
I think this should stay maxNameLength(), because it's directly creating the stat name

ggreenway commented on test/server/hot_restart_impl_test.cc:
```diff
@@ -132,15 +136,15 @@ TEST_P(HotRestartImplAlignmentTest, objectOverlap) {
     // If this isn't true then the hard coded part of the name isn't long enough to make the test
     // valid.
-    EXPECT_EQ(ts.name_.size(), Stats::RawStatData::maxNameLength());
+    EXPECT_EQ(ts.name_.size(), Stats::RawStatData::maxObjNameLength());
```
Same here: maxNameLength()

ggreenway commented on test/server/hot_restart_impl_test.cc:
```diff
@@ -50,8 +52,10 @@ class HotRestartImplTest : public testing::Test {
-                                         options_.maxStats(), options_.maxStatNameLength()));
+  EXPECT_EQ(hot_restart_->version(),
+            Envoy::Server::SharedMemory::version(options_.maxStats(),
+                                                 options_.maxObjNameLength() +
```
I think this is correct, but as per @mattklein123's request, can you make sure the version string is the same before and after your change (as a manual check)?

rshriram replied:
with the new build its `9.200.16384.127` and with master build, its `9.200.16384.127` . so it matches!

mattklein123 commented on bazel/README.md:
```diff
@@ -245,11 +245,12 @@ on the Bazel command line.
 
-The default maximum number of stats in shared memory, and the default maximum length of a stat name,
-can be overriden at compile-time by defining `ENVOY_DEFAULT_MAX_STATS` and `ENVOY_DEFAULT_MAX_STAT_NAME_LENGTH`,
+The default maximum number of stats in shared memory, and the default maximum length of a cluster/
```
nit: reflow text.

mattklein123 commented on source/common/config/json_utility.h:
```diff
@@ -6,6 +6,12 @@
+// Set a string field in a protobuf message with the corresponding string value
+#define JSON_UTIL_SET_STRING_VALUE(json, message, field_name, value)                               \
+  do {                                                                                             \
+    (message).set_##field_name(value);                                                             \
```
What does this have to do with JSON? Why is this macro needed at all?

mattklein123 commented on source/server/options_impl.cc:
```diff
@@ -82,9 +83,12 @@ OptionsImpl::OptionsImpl(int argc, char** argv, const HotRestartVersionCb& hot_r
-  TCLAP::ValueArg<uint64_t> max_stat_name_len("", "max-stat-name-len",
-                                              "Maximum name length for a stat", false,
-                                              ENVOY_DEFAULT_MAX_STAT_NAME_LENGTH, "uint64_t", cmd);
+  TCLAP::ValueArg<uint64_t> max_obj_name_len("", "max-obj-name-len",
```
needs doc update

rshriram replied:
ack. Sorry missed that totally.

mattklein123 commented on test/common/upstream/cluster_manager_impl_test.cc:
```diff
@@ -242,10 +242,7 @@ TEST_F(ClusterManagerImplTest, MaxClusterName) {
-                            "JSON at lines 4-6 does not conform to schema.\n Invalid schema: "
-                            "#/properties/name\n Schema violation: maxLength\n Offending "
-                            "document key: #/name");
+  EXPECT_THROW(create(parseBootstrapFromJson(json)), EnvoyException);
```
Why does this throw now? What's the new error?

rshriram replied:
This was always throwing an error. Earlier, it threw a JSON schema error with some message like "Invalid #/properties/name...." .. I merely changed this to use EnvoyException, as we are no longer validating the length using JSON schema.

In other words, UX wise, you would still get an exception when max length is exceeded, except it will be worded differently.

mattklein123 replied:
Right, please keep EXPECT_THROW_WITH_MESSAGE and switch it to the new error.

hennna commented on source/common/stats/stats_impl.h:
```diff
@@ -103,6 +115,8 @@ struct RawStatData {
 
 private:
   static const size_t DEFAULT_MAX_NAME_SIZE = 127;
```
Should this be DEFAULT_MAX_OBJ_NAME_SIZE? Wondering if it would be better to have all max name size parameters be derived.

hennna commented on source/server/options_impl.cc:
```diff
@@ -18,12 +19,12 @@
-#if ENVOY_DEFAULT_MAX_STAT_NAME_LENGTH < 127
-#error "ENVOY_DEFAULT_MAX_STAT_NAME_LENGTH must be >= 127"
+#if ENVOY_DEFAULT_MAX_OBJ_NAME_LENGTH < 60
+#error "ENVOY_DEFAULT_MAX_OBJ_NAME_LENGTH must be >= 60"
```
Outside the scope of this PR... Will there come a time when we also want to have MAX_STAT_SUFFIX_LENGTH be configurable? Or is that better left hard coded and bumped up when necessary?

rshriram commented:
On Tue, Oct 17, 2017 at 11:08 PM hennna <notifications@github.com> wrote:

> *@hennna* commented on this pull request.
>
> Thanks! Looking forward to using this feature.
> ------------------------------
>
> In source/common/stats/stats_impl.h
> <https://github.com/envoyproxy/envoy/pull/1871#discussion_r145274398>:
>
> > @@ -103,6 +115,8 @@ struct RawStatData {
>
>  private:
>    static const size_t DEFAULT_MAX_NAME_SIZE = 127;
> +  // total length must be equal to DEFAULT_MAX_NAME_SIZE
> +  static const size_t MAX_STAT_SUFFIX_LENGTH = 67;
>
> Maybe document where this number comes from? It seems from #667
> <https://github.com/envoyproxy/envoy/pull/667> it was from a cluster
> stats name plus some buffer.
> ------------------------------
>
> In source/common/stats/stats_impl.h
> <https://github.com/envoyproxy/envoy/pull/1871#discussion_r145306685>:
>
> > @@ -68,6 +68,18 @@ struct RawStatData {
>    }
>
>    /**
> +   * Returns the maximum length of a user supplied object (route/cluster/listener)
> +   * name field in a stat. This length does not include a trailing NULL-terminator.
> +   */
> +  static size_t maxObjNameLength() { return maxNameLength() - MAX_STAT_SUFFIX_LENGTH; }
>
> Should we instead have maxNameLength be derived from maxObjNameLength?
>

We could. I was trying to avoid touching the functions that were playing
tricks with static vars, as the maxNameLen function is used everywhere! But
let me see if it’s easy to change.

> ------------------------------
>
> In source/common/stats/stats_impl.h
> <https://github.com/envoyproxy/envoy/pull/1871#discussion_r145306841>:
>
> > @@ -103,6 +115,8 @@ struct RawStatData {
>
>  private:
>    static const size_t DEFAULT_MAX_NAME_SIZE = 127;
>
> Should this be DEFAULT_MAX_OBJ_NAME_SIZE? Wondering if it would be better
> to have all max name size parameters be derived.
>

This is max length of a stat entry. I agree it’s worded in a bit confusing
manner.

> ------------------------------
>
> In source/server/options_impl.cc
> <https://github.com/envoyproxy/envoy/pull/1871#discussion_r145307198>:
>
> >  #endif
>
> -#if ENVOY_DEFAULT_MAX_STAT_NAME_LENGTH < 127
> -#error "ENVOY_DEFAULT_MAX_STAT_NAME_LENGTH must be >= 127"
> +#if ENVOY_DEFAULT_MAX_OBJ_NAME_LENGTH < 60
> +#error "ENVOY_DEFAULT_MAX_OBJ_NAME_LENGTH must be >= 60"
>
> Outside the scope of this PR... Will there come a time when we also want
> to have MAX_STAT_SUFFIX_LENGTH be configurable? Or is that better left hard
> coded and bumped up when necessary?
>

@mattklein123 says we can bump it when necessary. I am surprised that 67
characters are enough to hold our current stat names.

> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/envoyproxy/envoy/pull/1871#pullrequestreview-70045448>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AH0qdwZDap0dY9z28VKZ1rTKU3HTS0Qlks5stWuzgaJpZM4P7TBJ>
> .
>

rshriram commented:
@mattklein123 / @junr03 I noticed this in the docs: 

```Virtual cluster statistics are output in the
*vhost.<virtual host name>.vcluster.<virtual cluster name>.* namespace and include the following
statistics:
```

Given that this PR allows for configurable name lengths for clusters, how does it affect the above stat, which is concatenating two different user supplied identities?

mattklein123 commented:
> Given that this PR allows for configurable name lengths for clusters, how does it affect the above stat, which is concatenating two different user supplied identities?

This occurred to me also. I think there are probably a bunch of random places where we should be guarding against user supplied stat names, etc. that might be too large. I would probably just open up a follow up ticket to audit.

rshriram commented:
@mattklein123 I moved all the objNameLength checks into a separate file. Also added name length checks for virtual host (but not the virtual host + virtual cluster). 
@hennna addressed your comments as well. PTAL

mattklein123 commented on docs/operations/cli.rst:
```diff
@@ -117,10 +117,19 @@ following are the command line options that Envoy supports.
   information. Defaults to 900 seconds (15 minutes).
 
-.. option:: --max-stat-name-len <uint64_t>
+.. _operations_cli_max_obj_name_len:
```
doc nit: You don't need this anchor, you can reference options directly and then they look nice inline. See how we ref other options.

mattklein123 commented on bazel/README.md:
```diff
@@ -245,11 +245,14 @@ on the Bazel command line.
-can be overriden at compile-time by defining `ENVOY_DEFAULT_MAX_STATS` and `ENVOY_DEFAULT_MAX_STAT_NAME_LENGTH`,
-respectively, to the desired value.  For example:
+The default maximum number of stats in shared memory, and the default
+maximum length of a cluster/ route config/listener name, can be
```
nit: extra space in "cluster/ "

mattklein123 commented on test/common/upstream/cluster_manager_impl_test.cc:
```diff
@@ -232,22 +232,6 @@ TEST_F(ClusterManagerImplTest, UnknownHcType) {
-  }
-  )EOF";
-
-  EXPECT_THROW_WITH_MESSAGE(create(parseBootstrapFromJson(json)), Json::Exception,
```
Isn't this test still valid? Or is it not anymore?

rshriram replied:
Same comment as above. Test just moved to utility_test.cc ..

mattklein123 replied:
Yup sounds good

mattklein123 commented on test/common/router/rds_impl_test.cc:
```diff
@@ -182,6 +182,27 @@ TEST_F(RdsImplTest, UnknownCluster) {
+  }
+  )EOF";
+
+  EXPECT_THROW(RouteConfigProviderUtil::create(parseHttpConnectionManagerFromJson(config_json),
```
Can we do EXPECT_THROW_WITH_MESSAGE here?

rshriram replied:
Actually, can I get rid of it? We have the same test in utility_test.cc

mattklein123 replied:
Yup that's fine

ggreenway commented on source/server/options_impl.cc:
```diff
@@ -93,14 +99,16 @@ OptionsImpl::OptionsImpl(int argc, char** argv, const HotRestartVersionCb& hot_r
   if (hot_restart_version_option.getValue()) {
-    std::cerr << hot_restart_version_cb(max_stats.getValue(), max_stat_name_len.getValue());
+    std::cerr << hot_restart_version_cb(max_stats.getValue(),
+                                        max_obj_name_len.getValue() +
```
It would be nicer if we could call RawStatData::configure(), then use RawStatData::maxNameLength() here, instead of doing the addition.  But I think it's probably more effort than it's worth, because the options_impl doesn't have it's values yet at this point in the function, so probably fine to leave as is, and the tests should catch if this gets messed up somehow.

hennna commented:
LGTM!
